File: 814-binary-tree-pruning/cpp/main.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
 public:
  TreeNode* pruneTree(TreeNode* root) {
    if (containsOneAndPrune(root)) {
      return root;
    }
    return NULL;
  }

  bool containsOneAndPrune(TreeNode* root) {
    if (root == NULL) {
      return false;
    }
    bool contains = root->val == 1;
    bool left = containsOneAndPrune(root->left);
    bool right = containsOneAndPrune(root->right);

    if (!left) {
      root->left = NULL;
    }

    if (!right) {
      root->right = NULL;
    }

    return contains || left || right;
  }
};

File: 814-binary-tree-pruning/go/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func pruneTree(root *TreeNode) *TreeNode {
	if pruneTreeAndReturn(root) {
		return root
	}
	return nil
}

func pruneTreeAndReturn(root *TreeNode) bool {
	if root == nil {
		return false
	}

	self := root.Val == 1
	left := pruneTreeAndReturn(root.Left)
	right := pruneTreeAndReturn(root.Right)

	if !left {
		root.Left = nil
	}

	if !right {
		root.Right = nil
	}

	return self || left || right
}
